File: model/src/main/java/no/nav/pensjon/vtp/testmodell/identer/LokalIdentIndeks.java
package no.nav.pensjon.vtp.testmodell.identer;

import java.util.Collections;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

import no.nav.pensjon.vtp.testmodell.kodeverk.Kjønn;

/** konverterer lokale identer brukt i testcase til utvalgte fødselsnummer hentet fra syntetisk liste. */
public class LokalIdentIndeks {
    private final IdentGenerator identGenerator;
    private final Map<String, String> identer = new ConcurrentHashMap<>(); // NOSONAR
    private final String unikScenarioId;

    public LokalIdentIndeks(String unikScenarioId, IdentGenerator identGenerator) {
        this.unikScenarioId = unikScenarioId;
        this.identGenerator = identGenerator;
    }

    public Map<String, String> getAlleIdenter(){
        return Collections.unmodifiableMap(identer);
    }

    public String getVoksenIdentForLokalIdent(String lokalIdent, Kjønn kjønn) {
        if (lokalIdent.matches("^\\d+$")) {
            return identer.computeIfAbsent(key(lokalIdent), i -> lokalIdent);
        }
        return identer.computeIfAbsent(key(lokalIdent), i -> kjønn == Kjønn.M ? identGenerator.tilfeldigMannFnr(null) : identGenerator.tilfeldigKvinneFnr(null));
    }

    public String getVoksenIdentForLokalIdent(String lokalIdent, Kjønn kjønn, String foedselsdato) {
        if (lokalIdent.matches("^\\d+$")) {
            return identer.computeIfAbsent(key(lokalIdent), i -> lokalIdent);
        }
        return identer.computeIfAbsent(key(lokalIdent), i -> kjønn == Kjønn.M ? identGenerator.tilfeldigMannFnr(foedselsdato) : identGenerator.tilfeldigKvinneFnr(foedselsdato));
    }

    public String getVoksenIdentForLokalIdent(String lokalIdent) {
        if (lokalIdent.matches("^\\d+$")) {
            return identer.computeIfAbsent(key(lokalIdent), i -> lokalIdent);
        }
        return identer.computeIfAbsent(key(lokalIdent), i -> identGenerator.tilfeldigFnr());
    }

    private String key(String lokalIdent) {
        return unikScenarioId + "::" + lokalIdent;
    }

    public String getBarnIdentForLokalIdent(String lokalIdent) {
        if (lokalIdent.matches("^\\d+$")) {
            return identer.computeIfAbsent(key(lokalIdent), i -> lokalIdent);
        }
        // tilfeldig kjønn
        return identer.computeIfAbsent(key(lokalIdent), i -> identGenerator.tilfeldigBarnUnderTreAarFnr());
    }

    public String getIdent(String lokalIdent) {
        String key = key(lokalIdent);
        String ident = identer.get(key);
//        if (ident == null) {
//            throw new IllegalStateException("Kjenner ikke ident for lokal id: " + lokalIdent + ", mulig ikke lastet ennå?");
//        }
        return ident;
    }
}

File: core/src/main/kotlin/no/nav/pensjon/vtp/testmodell/identer/LokalIdentIndeks.kt
package no.nav.pensjon.vtp.testmodell.identer

import no.nav.pensjon.vtp.testmodell.kodeverk.Kjønn
import java.util.concurrent.ConcurrentHashMap

/** konverterer lokale identer brukt i testcase til utvalgte fødselsnummer hentet fra syntetisk liste.  */
class LokalIdentIndeks(private val unikScenarioId: String, private val identGenerator: IdentGenerator) {
    private val identer: MutableMap<String, String> = ConcurrentHashMap()

    fun getVoksenIdentForLokalIdent(lokalIdent: String, kjønn: Kjønn): String {
        return if (lokalIdent.matches("^\\d+$".toRegex())) {
            identer.computeIfAbsent(key(lokalIdent)) { lokalIdent }
        } else identer.computeIfAbsent(key(lokalIdent)) {
            if (kjønn == Kjønn.M) identGenerator.tilfeldigMannFnr(
                null
            ) else identGenerator.tilfeldigKvinneFnr(null)
        }
    }

    fun getVoksenIdentForLokalIdent(lokalIdent: String, kjønn: Kjønn, foedselsdato: String?): String {
        return if (lokalIdent.matches("^\\d+$".toRegex())) {
            identer.computeIfAbsent(key(lokalIdent)) { lokalIdent }
        } else identer.computeIfAbsent(key(lokalIdent)) {
            if (kjønn == Kjønn.M) identGenerator.tilfeldigMannFnr(
                foedselsdato
            ) else identGenerator.tilfeldigKvinneFnr(foedselsdato)
        }
    }

    private fun key(lokalIdent: String): String {
        return "$unikScenarioId::$lokalIdent"
    }

    fun getIdent(lokalIdent: String): String? {
        return identer[key(lokalIdent)]
    }
}
